src/engine/IdTable.h

```diff
@@ -0,0 +1,933 @@
+                init.begin(), sizeof(Id) * IdTableImpl<COLS>::_cols);
+    IdTableImpl<COLS>::_size++;
+  }
+
+  /**
+   * @brief Read cols() elements from init and stores them in a new row
+   **/
+  void push_back(const Id* init) {
+    if (IdTableImpl<COLS>::_size + 1 >= IdTableImpl<COLS>::_capacity) {
+      grow();
+    }
+    std::memcpy(IdTableImpl<COLS>::_data +
+                    IdTableImpl<COLS>::_size * IdTableImpl<COLS>::_cols,
+                init, sizeof(Id) * IdTableImpl<COLS>::_cols);
+    IdTableImpl<COLS>::_size++;
+  }
+
+  /**
+   * @brief Read cols() elements from init and stores them in a new row
+   **/
+  void push_back(const Row& init) {
+    assert(init.cols() == cols());
+    if (IdTableImpl<COLS>::_size + 1 >= IdTableImpl<COLS>::_capacity) {
+      grow();
+    }
+    std::memcpy(IdTableImpl<COLS>::_data +
+                    IdTableImpl<COLS>::_size * IdTableImpl<COLS>::_cols,
+                init.data(), sizeof(Id) * IdTableImpl<COLS>::_cols);
+    IdTableImpl<COLS>::_size++;
+  }
+
+  /**
+   * @brief Read cols() elements from init and stores them in a new row
+   **/
+  void push_back(ConstRow init) {
+    if (IdTableImpl<COLS>::_size + 1 >= IdTableImpl<COLS>::_capacity) {
+      grow();
+    }
+    std::memcpy(IdTableImpl<COLS>::_data +
+                    IdTableImpl<COLS>::_size * IdTableImpl<COLS>::_cols,
+                init.data(), sizeof(Id) * IdTableImpl<COLS>::_cols);
+    IdTableImpl<COLS>::_size++;
+  }
+
+  void push_back(const IdTableStatic<COLS>& init, size_t row) {
+    assert(init.IdTableImpl<COLS>::_cols == IdTableImpl<COLS>::_cols);
+    if (IdTableImpl<COLS>::_size + 1 >= IdTableImpl<COLS>::_capacity) {
+      grow();
+    }
+    std::memcpy(IdTableImpl<COLS>::_data +
+                    IdTableImpl<COLS>::_size * IdTableImpl<COLS>::_cols,
+                init.IdTableImpl<COLS>::_data + row * IdTableImpl<COLS>::_cols,
+                sizeof(Id) * IdTableImpl<COLS>::_cols);
+    IdTableImpl<COLS>::_size++;
+  }
+
+  void pop_back() {
+    if (IdTableImpl<COLS>::_size > 0) {
+      IdTableImpl<COLS>::_size--;
+    }
+  }
+
+  /**
+   * @brief Inserts the elements in the range [begin;end) before pos.
+   **/
+  void insert(const iterator& pos, const iterator& begin, const iterator& end) {
+    assert(begin.cols() == cols());
+    if (begin.row() >= end.row()) {
+      return;
+    }
+    size_t target = std::min(pos.row(), this->end().row());
+    size_t numNewRows = end.row() - begin.row();
+    if (IdTableImpl<COLS>::_capacity < IdTableImpl<COLS>::_size + numNewRows) {
+      size_t numMissing =
+          IdTableImpl<COLS>::_size + numNewRows - IdTableImpl<COLS>::_capacity;
+      grow(numMissing);
+    }
+    // Move the data currently in the way back. As the src and target
+    // of the copy might overlap the rows are copied individually starting
+    // at the back. The loop still counts up to avoid underflows.
```

Comment:
I think `memmove` allows for overlapping memory, could we use this?

Reply:
I swapped it out for memmove and added tests for inserting into IdTables.